pulsar-functions/proto/src/main/proto/InstanceCommunication.proto

```diff
@@ -72,5 +72,7 @@ message HealthCheckResult {
 service InstanceControl {
     rpc GetFunctionStatus(google.protobuf.Empty) returns (FunctionStatus) {}
     rpc GetAndResetMetrics(google.protobuf.Empty) returns (MetricsData) {}
+    rpc ResetMetrics(google.protobuf.Empty) returns (google.protobuf.Empty) {}
```

Comment:
Get and Reset always go hand in hand. I believe thats the fundamental issue with this pr.

Reply:
actually, getAndReset doesn't store any sample because it gets current sample and resets it. right now, it is only used by prometheus to feed into metrics. but if we want to fetch metrics on demand then we want to update metrics (reset) periodically and get metrics on demand. PR #2130 in created on top of this one which aggregates metrics into function stats so, we can always see function metrics on demand using getMetrics api similar like we can get for topic: `pulsar-admin persistent stats <topic>` which gives current topic stats using cli api. also, we can add additional REST api which can generate metrics in different format other than prometheus.

Reply:
> It goes against established compute patterns. This task of keeping track in a durable way of the metrics is usually what the external systems(like prometheus) do.

I think there is a misunderstanding here. This will not keep durable accumulated metrics but it only gives transient metrics value (last minute metrics) for a function stored at function. 
In this PR, each function will keep only one sample of the last minute, there is a schedule task which updates the metrics of each function every minute and function keeps one sample with it (same thing happens in pulsar, bk, zk where it capture last minute of metrics). So, whenever anyone requires a metrics for a function it can always give latest sample of metrics for that function. So, prometheus and any other monitoring system can call this api at every minute and can get last minute metrics. 
Therefore, with #2130 `FunctionStats` also contains function-metrics in it so, as a admin if I want to check current metrics and stats of a function then I can always get it using api/cli same way we already has in pulsar `stats and stats-internal` and this api is essentially part of admin-monitoring.
with existing implementation, `getAndRest` triggered by prometheus (or any trigger-point) which gets metrics and after that there is no way to get last minute of metrics and there is no way for admin to see current metrics of the function. 
I hope this make sense.